File: platform-domain/src/main/java/com/platform/domain/model/AuditEntry.java
package com.platform.domain.model;

import java.time.Instant;
import java.util.List;

/**
 * Immutable audit log entry. Written once by domain services, never updated.
 * changedFields values are UNMASKED — the audit log is the authoritative unmasked record.
 * Access is controlled by audit query roles.
 */
public record AuditEntry(
        String id,
        String tenantId,
        String workItemId,
        String correlationId,
        AuditEventType eventType,
        String previousState,
        String newState,
        String transitionName,
        List<ChangedField> changedFields,
        String actorUserId,
        String actorRole,
        Instant timestamp,
        String idempotencyKey
) {
    public record ChangedField(String fieldPath, Object previousValue, Object newValue) {}
}

File: platform-frontend/src/types/AuditEntry.ts
import { z } from 'zod';

export const AuditEventTypeSchema = z.enum([
  'INGESTION', 'STATE_TRANSITION', 'FIELD_UPDATE', 'ASSIGNMENT',
  'ROUTING_FALLBACK', 'MAKER_CHECKER_APPROVAL', 'GROUP_REASSIGNMENT',
  'CHILD_WORKFLOW_CREATED', 'ATTACHED_TO_EXISTING_CHILD', 'DEPENDENCY_RESOLVED',
  'SLA_WARNING', 'SLA_BREACH', 'DUPLICATE_INGESTION_DISCARDED', 'TRANSITION_FAILED',
]);

export const ChangedFieldSchema = z.object({
  fieldPath: z.string(),
  previousValue: z.unknown(),
  newValue: z.unknown(),
});

export const AuditEntrySchema = z.object({
  id: z.string(),
  tenantId: z.string(),
  workItemId: z.string(),
  correlationId: z.string(),
  eventType: AuditEventTypeSchema,
  previousState: z.string().nullable(),
  newState: z.string().nullable(),
  transitionName: z.string().nullable(),
  changedFields: z.array(ChangedFieldSchema),
  actorUserId: z.string(),
  actorRole: z.string(),
  timestamp: z.string(),
  idempotencyKey: z.string().nullable(),
});

export type AuditEntry = z.infer<typeof AuditEntrySchema>;
